### PR TITLE
ArC - fix handling of disposed parameter qualifiers

### LINKED ISSUE #1839: NPE in io.quarkus:quarkus-maven-plugin:0.12.0:build
`quarkus-maven-plugin:build` fails with a `NullPointerException`:

````
[ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:0.12.0:build (default) on project kaiserkai-rest-impl: Failed to build a runnable JAR: Failed to build a runner jar: Failed to augment application classes: Build failure: Build failed due to errors
[ERROR]         [error]: Build step io.quarkus.arc.deployment.ArcAnnotationProcessor#build threw an exception: java.lang.NullPointerException
````
The detailed stack trace is:

````
Caused by: java.lang.NullPointerException
    at io.quarkus.arc.processor.Beans.hasQualifier (Beans.java:405)
    at io.quarkus.arc.processor.BeanDeployment.findDisposer (BeanDeployment.java:694)
    at io.quarkus.arc.processor.BeanDeployment.findBeans (BeanDeployment.java:635)
    at io.quarkus.arc.processor.BeanDeployment.<init> (BeanDeployment.java:136)
    at io.quarkus.arc.processor.BeanProcessor.process (BeanProcessor.java:150)
    at io.quarkus.arc.deployment.ArcAnnotationProcessor.build (ArcAnnotationProcessor.java:237)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at io.quarkus.deployment.ExtensionLoader$1.execute (ExtensionLoader.java:506)
    at org.jboss.builder.BuildContext.run (BuildContext.java:413)
    at org.jboss.threads.ContextClassLoaderSavingRunnable.run (ContextClassLoaderSavingRunnable.java:35)
    at org.jboss.threads.EnhancedQueueExecutor.safeRun (EnhancedQueueExecutor.java:1998)
    at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask (EnhancedQueueExecutor.java:1525)
    at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run (EnhancedQueueExecutor.java:1416)
    at java.lang.Thread.run (Thread.java:748)
    at org.jboss.threads.JBossThread.run (JBossThread.java:479)

````

I don't fully understand what's going on here, but there seems to be a broken API contract: `io.quarkus.arc.processor.Beans.hasQualifier(ClassInfo, AnnotationInstance, Collection<AnnotationInstance>)` expects the third argument to be non-null, whereas `BeanDeployment.findDisposer()` calls it with a null value.

### PR DESCRIPTION
**Pull request by mkouba**
- resolves #1839

### CODE REVIEW

**Comment by mkouba** [comment]
@hwellmann Could you pls try this patch with your application?

**Comment by hwellmann** [comment]
Testing with a local build of your branch, the original NPE no longer occurs, but now I'm getting a similar one related to producers:

````
Caused by: java.lang.NullPointerException
    at io.quarkus.arc.processor.Types.getTypeClosure (Types.java:245)
    at io.quarkus.arc.processor.Types.getTypeClosure (Types.java:160)
    at io.quarkus.arc.processor.Beans.createProducerMethod (Beans.java:118)
    at io.quarkus.arc.processor.BeanDeployment.findBeans (BeanDeployment.java:634)
    at io.quarkus.arc.processor.BeanDeployment.<init> (BeanDeployment.java:136)
    at io.quarkus.arc.processor.BeanProcessor.process (BeanProcessor.java:150)
    at io.quarkus.arc.deployment.ArcAnnotationProcessor.build (ArcAnnotationProcessor.java:259)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at io.quarkus.deployment.ExtensionLoader$1.execute (ExtensionLoader.java:506)
    at org.jboss.builder.BuildContext.run (BuildContext.java:413)
    at org.jboss.threads.ContextClassLoaderSavingRunnable.run (ContextClassLoaderSavingRunnable.java:35)
    at org.jboss.threads.EnhancedQueueExecutor.safeRun (EnhancedQueueExecutor.java:1998)
    at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask (EnhancedQueueExecutor.java:1525)
    at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run (EnhancedQueueExecutor.java:1416)
    at java.lang.Thread.run (Thread.java:748)
    at org.jboss.threads.JBossThread.run (JBossThread.java:479)
````

**Review by manovotn — APPROVED** [review]
Looks good!

**Comment by mkouba** [comment]
@hwellmann Thanks. The other one is quite weird. Is your producer annotated with `@Typed()`?

**Comment by hwellmann** [comment]
Almost: the class of the producer's return type is annotated `@Typed()` with no arguments.

(The producer method is `org.apache.deltaspike.core.util.ProjectStageProducer.getProjectStage()`).

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/DisposerInfo.java [hunk 1]
```diff
@@ -73,6 +76,16 @@ void init(List<Throwable> errors) {
         }
     }
 
+    Collection<AnnotationInstance> getDisposedParameteterQualifiers() {
```
Comment (hwellmann) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/DisposerInfo.java]: Parameteter -> Parameter

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/DisposerInfo.java [hunk 2]
```diff
@@ -73,6 +76,16 @@ void init(List<Throwable> errors) {
...
+        return Annotations.getParameterAnnotations(declaringBean.getDeployment(), disposerMethod, disposedParameter.position())
+                .stream().filter(a -> declaringBean.getDeployment().getQualifier(a.name()) != null)
+                .collect(Collectors.toList());
+    }
+
+    Type getDiposedParameterType() {
```
Comment (hwellmann) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/DisposerInfo.java]: Diposed -> Disposed

**Comment by mkouba** [comment]
@hwellmann Ok, I'll try to fix this case. But note that deltaspike will very likely not work on quarkus - it heavily relies on portable extensions that are not supported.

**Comment by hwellmann** [comment]
Yes, I know that DeltaSpike won't work. I'm converting an existing project, and the first shot was just to add `quarkus-maven-plugin` and run it. At least it seems to help discover some edge cases ;-)

**Comment by mkouba** [comment]
@hwellmann @manovotn Ok, one more fix added. If CI is green and you're satisfied we should merge this PR and eventually create a new issue (I do expect more corner cases like this ;-).

**Comment by hwellmann** [comment]
Test successful. The stack trace no longer occurs.

**Comment by mkouba** [comment]
@hwellmann Thanks!

**Review by manovotn — APPROVED** [review]

**Pull request merged**